tests/end2end/runner.sh

```diff
@@ -4,21 +4,46 @@ source ./utils.sh
 
 # Create an associated array mapping test name to exit code.
 declare -A TESTS
+# Map testname to the number of times the test was run.
+declare -A RUNS
 
+TRIALS=3
+
+# Each sub-directory in the current directory is a test.
 for test in ./*/
 do
+    # Remove things like dir / from the test name
     testname="${test%*/}"
     testname="${testname#./}"
     yellow_echo "Running Test: ${testname}"
-    echo "============================================================"
+    echo "================================================================="
+    # Move into the test dir
     pushd ${testname}
-    if [[ -f ./clean.sh ]]; then
-        ./clean.sh
-    else
-        ../clean.sh
-    fi
-    ./test.sh
-    TESTS[${testname}]=${?}
+    # Run the test up to ${TRIALS} times, stopping when it has a return of 0
+    i=0
+    # This is just a non-zero value, just to get us into the loop, it doesn't
+    # matter what the value is as it will be overwritten by the test return code.
+    return_code=2
```

Comment:
Total nit but maybe it would be cleaner to make it an impossible return code like -1